File: scripts/sec-insider-scanner.ts
#!/usr/bin/env npx ts-node
/**
 * SEC Form 4 Insider Buy Scanner
 *
 * Fetches Form 4 filings from SEC EDGAR for the last 24 hours,
 * filters for open-market purchases > $100K, and posts a ranked
 * summary to Slack.
 *
 * Usage:
 *   npx ts-node scripts/sec-insider-scanner.ts
 *   npx ts-node scripts/sec-insider-scanner.ts --dry-run   # print only, no Slack post
 *   npx ts-node scripts/sec-insider-scanner.ts --hours 48  # look back further
 *
 * Environment variables:
 *   SEC_CONTACT_EMAIL   – required by SEC fair-use policy (your email)
 *   SLACK_WEBHOOK_URL   – Slack Incoming Webhook URL  OR
 *   SLACK_BOT_TOKEN     – Slack Bot token (needs chat:write scope)
 *   SLACK_CHANNEL_ID    – channel to post to (default: #sec-form4-insider-scanner)
 *   MIN_PURCHASE_VALUE  – minimum USD value to include (default: 100000)
 *
 * Cron example (run every weekday at 6 AM EST):
 *   0 11 * * 1-5 cd /path/to/project && npx ts-node scripts/sec-insider-scanner.ts >> /var/log/insider-scanner.log 2>&1
 */

import * as https from 'https'
import * as http from 'http'

// ─── Config ───────────────────────────────────────────────────────────────────

const SEC_UA = `TradingJournalPro ${process.env.SEC_CONTACT_EMAIL ?? 'support@tradingjournalpro.com'}`
const SLACK_WEBHOOK_URL = process.env.SLACK_WEBHOOK_URL ?? ''
const SLACK_BOT_TOKEN = process.env.SLACK_BOT_TOKEN ?? ''
const SLACK_CHANNEL_ID = process.env.SLACK_CHANNEL_ID ?? 'C0AUARBCPND'
const MIN_BUY_VALUE = Number(process.env.MIN_PURCHASE_VALUE ?? 100_000)

const args = process.argv.slice(2)
const DRY_RUN = args.includes('--dry-run')
const HOURS_BACK = Number(args[args.indexOf('--hours') + 1] ?? 24)

// ─── HTTP helper (no external deps) ──────────────────────────────────────────

function get(url: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const mod = url.startsWith('https') ? https : http
    const req = mod.get(url, { headers: { 'User-Agent': SEC_UA, Accept: '*/*' } }, res => {
      if (res.statusCode === 301 || res.statusCode === 302) {
        resolve(get(res.headers.location!))
        return
      }
      if (res.statusCode !== 200) {
        reject(new Error(`HTTP ${res.statusCode}: ${url}`))
        return
      }
      let data = ''
      res.on('data', (chunk: string) => (data += chunk))
      res.on('end', () => resolve(data))
    })
    req.on('error', reject)
    req.setTimeout(15_000, () => { req.destroy(); reject(new Error(`Timeout: ${url}`)) })
  })
}

function post(url: string, body: string): Promise<string> {
  return new Promise((resolve, reject) => {
    const parsed = new URL(url)
    const options = {
      hostname: parsed.hostname,
      path: parsed.pathname + parsed.search,
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        'Content-Length': Buffer.byteLength(body),
        ...(SLACK_BOT_TOKEN ? { Authorization: `Bearer ${SLACK_BOT_TOKEN}` } : {}),
      },
    }
    const mod = parsed.protocol === 'https:' ? https : http
    const req = mod.request(options, res => {
      let data = ''
      res.on('data', (c: string) => (data += c))
      res.on('end', () => resolve(data))
    })
    req.on('error', reject)
    req.write(body)
    req.end()
  })
}

// ─── Types ────────────────────────────────────────────────────────────────────

interface InsiderBuy {
  rank: number
  company: string
  ticker: string
  insiderName: string
  title: string
  shares: number
  pricePerShare: number
  totalValue: number
  transactionDate: string
  filingUrl: string
}

// ─── EDGAR data pipeline ──────────────────────────────────────────────────────

async function getRecentFilings(startDate: string, endDate: string): Promise<string[]> {
  const url =
    `https://efts.sec.gov/LATEST/search-index?forms=4` +
    `&dateRange=custom&startdt=${startDate}&enddt=${endDate}` +
    `&hits.hits._source=accession_no`

  const json = JSON.parse(await get(url))
  const hits: { _source: { accession_no: string } }[] = json.hits?.hits ?? []
  return hits.map(h => h._source.accession_no)
}

function extractTag(xml: string, tag: string): string {
  return xml.match(new RegExp(`<${tag}>(?:<value>)?([^<]+)(?:<\\/value>)?<\\/${tag}>`))?.[1]?.trim() ?? ''
}

async function parseForm4(accNo: string): Promise<InsiderBuy[]> {
  const cik = accNo.split('-')[0].replace(/^0+/, '')
  const accNoClean = accNo.replace(/-/g, '')
  const base = `https://www.sec.gov/Archives/edgar/data/${cik}/${accNoClean}`

  // Find primary XML filename from filing index
  let xmlFile = `${accNo}.xml`
  try {
    const index = await get(`${base}/${accNo}-index.htm`)
    const m = index.match(/href="([^"]*\.xml)"/i)
    if (m) xmlFile = m[1].split('/').pop()!
  } catch { /* use default */ }

  const xml = await get(`${base}/${xmlFile}`)

  const company = extractTag(xml, 'issuerName')
  const ticker = extractTag(xml, 'issuerTradingSymbol')
  const insiderName = extractTag(xml, 'rptOwnerName')
  const officerTitle = extractTag(xml, 'officerTitle')
  const isDirector = xml.includes('<isDirector>1</isDirector>')
  const isOfficer = xml.includes('<isOfficer>1</isOfficer>')
  const title = officerTitle || (isDirector ? 'Director' : isOfficer ? 'Officer' : 'Insider')

  const blocks = xml.match(/<nonDerivativeTransaction>[\s\S]*?<\/nonDerivativeTransaction>/g) ?? []
  const results: InsiderBuy[] = []

  for (const block of blocks) {
    const code = extractTag(block, 'transactionCode')
    const acqDisp = block.match(/<transactionAcquiredDisposedCode>\s*<value>([^<]+)<\/value>/)?.[1]?.trim()
    if (code !== 'P' && acqDisp !== 'A') continue

    const shares = parseFloat(block.match(/<transactionShares>\s*<value>([^<]+)<\/value>/)?.[1] ?? '0')
    const price = parseFloat(block.match(/<transactionPricePerShare>\s*<value>([^<]+)<\/value>/)?.[1] ?? '0')
    const date = block.match(/<transactionDate>\s*<value>([^<]+)<\/value>/)?.[1]?.trim() ?? ''
    const totalValue = shares * price

    if (totalValue < MIN_BUY_VALUE || shares <= 0 || price <= 0) continue

    results.push({ rank: 0, company, ticker, insiderName, title, shares, pricePerShare: price, totalValue, transactionDate: date, filingUrl: `${base}/` })
  }
  return results
}

function sleep(ms: number) {
  return new Promise(r => setTimeout(r, ms))
}

// ─── Slack payload ────────────────────────────────────────────────────────────

function fmt$(v: number) {
  return v >= 1_000_000 ? `$${(v / 1_000_000).toFixed(2)}M` : `$${Math.round(v / 1_000)}K`
}

function buildPayload(buys: InsiderBuy[], generatedAt: string) {
  const total = buys.reduce((s, b) => s + b.totalValue, 0)
  const rows = buys.slice(0, 15).map((b, i) => {
    const label = b.ticker ? `*${b.ticker}*  _${b.company}_` : `*${b.company}*`
    return (
      `*${i + 1}.* ${label}\n` +
      `   ${b.insiderName}  ·  ${b.title}\n` +
      `   ${fmt$(b.totalValue)}  ·  ${b.shares.toLocaleString()} sh @ $${b.pricePerShare.toFixed(2)}  ·  ${b.transactionDate}\n` +
      `   <${b.filingUrl}|View SEC Filing>`
    )
  })

  return {
    blocks: [
      { type: 'header', text: { type: 'plain_text', text: '📈 SEC Form 4 — Insider Purchases  |  Last 24 Hours', emoji: true } },
      {
        type: 'section',
        text: {
          type: 'mrkdwn',
          text:
            `*${buys.length} purchase${buys.length !== 1 ? 's' : ''}* above $100K  ·  Combined: *${fmt$(total)}*\n_${generatedAt}_`,
        },
      },
      { type: 'divider' },
      { type: 'section', text: { type: 'mrkdwn', text: rows.length ? rows.join('\n\n') : '_No qualifying purchases._' } },
      { type: 'divider' },
      {
        type: 'context',
        elements: [{ type: 'mrkdwn', text: 'Source: SEC EDGAR Form 4  ·  Ranked by total purchase value  ·  Directors & Officers only' }],
      },
    ],
  }
}

// ─── Main ─────────────────────────────────────────────────────────────────────

async function main() {
  const now = new Date()
  const from = new Date(now.getTime() - HOURS_BACK * 60 * 60 * 1000)
  const fmt = (d: Date) => d.toISOString().slice(0, 10)
  const generatedAt = now.toUTCString()

  console.log(`[scanner] Looking back ${HOURS_BACK}h · from ${fmt(from)} to ${fmt(now)}`)

  const accNos = await getRecentFilings(fmt(from), fmt(now))
  console.log(`[scanner] Found ${accNos.length} Form 4 filings`)

  const allBuys: InsiderBuy[] = []
  let parsed = 0

  for (let i = 0; i < accNos.length; i += 8) {
    const batch = accNos.slice(i, i + 8)
    const settled = await Promise.allSettled(batch.map(a => parseForm4(a)))
    for (const r of settled) {
      if (r.status === 'fulfilled') allBuys.push(...r.value)
      else console.warn('[scanner] parse error:', r.reason?.message)
    }
    parsed += batch.length
    if (i + 8 < accNos.length) await sleep(125) // respect SEC rate limit
  }

  allBuys.sort((a, b) => b.totalValue - a.totalValue)
  allBuys.forEach((b, i) => (b.rank = i + 1))

  console.log(`[scanner] Parsed ${parsed} filings · ${allBuys.length} buys > $${(MIN_BUY_VALUE / 1000).toFixed(0)}K`)
  allBuys.slice(0, 10).forEach(b =>
    console.log(`  #${b.rank} ${b.ticker || b.company} — ${b.insiderName} (${b.title}) — ${fmt$(b.totalValue)}`),
  )

  if (DRY_RUN) {
    console.log('[scanner] --dry-run: skipping Slack post')
    return
  }

  if (!SLACK_WEBHOOK_URL && !SLACK_BOT_TOKEN) {
    console.error('[scanner] Set SLACK_WEBHOOK_URL or SLACK_BOT_TOKEN to enable Slack posting')
    process.exit(1)
  }

  const payload = buildPayload(allBuys, generatedAt)

  if (SLACK_WEBHOOK_URL) {
    const result = await post(SLACK_WEBHOOK_URL, JSON.stringify(payload))
    console.log('[scanner] Webhook response:', result)
  } else {
    const body = JSON.stringify({ channel: SLACK_CHANNEL_ID, ...payload })
    const result = JSON.parse(await post('https://slack.com/api/chat.postMessage', body))
    if (!result.ok) throw new Error(`Slack error: ${result.error}`)
    console.log('[scanner] Posted to Slack:', result.ts)
  }
}

main().catch(err => {
  console.error('[scanner] Fatal:', err)
  process.exit(1)
})

File: scripts/sec_insider_scanner.py
#!/usr/bin/env python3
"""
SEC EDGAR Form 4 insider purchase scanner.

Scans Form 4 filings from the last 24 hours, filters for open-market purchases
by executives and directors exceeding $100 000, ranks by transaction value,
and posts a summary to a Slack channel.

Usage
-----
    # Print summary to stdout only
    python3 sec_insider_scanner.py

    # Also post to Slack (Incoming Webhook)
    SLACK_WEBHOOK_URL=https://hooks.slack.com/... python3 sec_insider_scanner.py

    # Custom look-back window and threshold
    python3 sec_insider_scanner.py --hours 48 --min-value 500000

Environment variables
---------------------
SLACK_WEBHOOK_URL   Slack Incoming Webhook URL (optional)
SLACK_CHANNEL       Override the default channel embedded in the webhook (optional)

SEC EDGAR API notes
-------------------
- Requires a descriptive User-Agent per https://www.sec.gov/privacy.htm#edgaraccess
- Rate limit: ~10 req/s; we stay well below that
"""

from __future__ import annotations

import argparse
import json
import os
import sys
import time
import xml.etree.ElementTree as ET
from datetime import datetime, timedelta, timezone
from urllib.error import HTTPError, URLError
from urllib.request import Request, urlopen

# ── Constants ────────────────────────────────────────────────────────────────

EDGAR_EFTS = "https://efts.sec.gov/LATEST/search-index"
EDGAR_ARCH = "https://www.sec.gov/Archives/edgar/data"
# SEC requires a descriptive User-Agent that includes contact info
USER_AGENT = "TradingJournalPro/1.0 (github.com/wilgut/trading-journal-pro; wilfred.gutierrez@gmail.com)"
HEADERS = {"User-Agent": USER_AGENT, "Accept": "application/json, text/xml, */*"}
DEFAULT_MIN_VALUE = 100_000   # $100 k
DEFAULT_HOURS = 24
MAX_FILINGS = 500             # Hard cap to avoid runaway loops
INTER_REQUEST_DELAY = 0.12    # seconds between EDGAR requests (~8 req/s)
TOP_N_SLACK = 25              # Maximum rows in the Slack table


# ── HTTP helpers ─────────────────────────────────────────────────────────────

def http_get(url: str, timeout: int = 20, retries: int = 3) -> bytes:
    for attempt in range(retries):
        try:
            req = Request(url, headers=HEADERS)
            with urlopen(req, timeout=timeout) as resp:
                return resp.read()
        except HTTPError as exc:
            if exc.code == 429:
                wait = 2 ** (attempt + 2)
                print(f"  Rate-limited; sleeping {wait}s …", file=sys.stderr)
                time.sleep(wait)
            elif attempt == retries - 1:
                raise
            else:
                time.sleep(2 ** attempt)
        except URLError:
            if attempt == retries - 1:
                raise
            time.sleep(2 ** attempt)
    raise RuntimeError(f"Failed to fetch {url}")  # unreachable


# ── EDGAR search ─────────────────────────────────────────────────────────────

def search_form4_filings(start: str, end: str) -> list[tuple[str, str]]:
    """
    Return list of (accession_number, cik) for Form 4 filings in [start, end].
    Both dates are ISO-8601 strings (YYYY-MM-DD).
    """
    results: list[tuple[str, str]] = []
    page_size = 100
    offset = 0

    while len(results) < MAX_FILINGS:
        url = (
            f"{EDGAR_EFTS}?forms=4"
            f"&dateRange=custom&startdt={start}&enddt={end}"
            f"&hits.hits.total=true&from={offset}"
        )
        raw = http_get(url)
        data = json.loads(raw)

        hits_block = data.get("hits", {})
        hits = hits_block.get("hits", [])
        if not hits:
            break

        for h in hits:
            acc_no = h.get("_id", "")
            src = h.get("_source", {})
            entity_ids = src.get("entity_id", [])
            if not entity_ids or not acc_no:
                continue
            cik = entity_ids[0] if isinstance(entity_ids, list) else entity_ids
            results.append((acc_no, str(cik)))

        total = hits_block.get("total", {}).get("value", 0)
        offset += page_size
        if offset >= total:
            break

    return results[:MAX_FILINGS]


# ── Filing XML retrieval ──────────────────────────────────────────────────────

def _filing_base_url(acc_no: str, cik: str) -> str:
    acc_clean = acc_no.replace("-", "")
    cik_int = str(int(cik))          # strip leading zeros
    return f"{EDGAR_ARCH}/{cik_int}/{acc_clean}"


def get_primary_xml(acc_no: str, cik: str) -> ET.Element | None:
    """
    Locate and return the root element of the primary Form 4 XML document.
    Tries the filing-index JSON first, then falls back to common name patterns.
    """
    base = _filing_base_url(acc_no, cik)

    # Strategy 1: use the filing index JSON to find the XML document
    try:
        idx = json.loads(http_get(f"{base}/{acc_no}-index.json"))
        for item in idx.get("directory", {}).get("item", []):
            name: str = item.get("name", "")
            # Primary Form 4 XML: ends with .xml, not an R-document, not a schema
            if (
                name.endswith(".xml")
                and not name.startswith("R")
                and "xsd" not in name.lower()
                and "xsl" not in name.lower()
                and "label" not in name.lower()
            ):
                raw = http_get(f"{base}/{name}")
                return ET.fromstring(raw)
    except Exception:
        pass

    # Strategy 2: common naming patterns
    acc_clean = acc_no.replace("-", "")
    for candidate in [f"{acc_clean}.xml", "form4.xml", "doc4.xml", "primaryDoc.xml"]:
        try:
            raw = http_get(f"{base}/{candidate}")
            return ET.fromstring(raw)
        except Exception:
            continue

    return None


# ── Transaction parsing ───────────────────────────────────────────────────────

def parse_purchases(root: ET.Element, min_value: float) -> list[dict]:
    """
    Extract open-market purchases by executives/directors from a Form 4 XML tree.
    Returns a list of purchase dicts sorted by value (highest first).
    """
    issuer_name = (root.findtext("issuer/issuerName") or "").strip()
    ticker = (root.findtext("issuer/issuerTradingSymbol") or "").strip().upper()

    owner_name = (
        root.findtext(".//reportingOwner/reportingOwnerId/rptOwnerName") or ""
    ).strip()

    rel = root.find(".//reportingOwner/reportingOwnerRelationship")
    title = ""
    is_exec_or_dir = False
    if rel is not None:
        title = (rel.findtext("officerTitle") or "").strip()
        is_dir = rel.findtext("isDirector") == "1"
        is_off = rel.findtext("isOfficer") == "1"
        is_exec_or_dir = is_dir or is_off
        if not title:
            if is_dir:
                title = "Director"
            elif is_off:
                title = "Officer"

    if not is_exec_or_dir:
        return []

    purchases: list[dict] = []
    for tx in root.findall(".//nonDerivativeTransaction"):
        # Transaction code "P" = open-market purchase
        code = tx.findtext(".//transactionCoding/transactionCode") or ""
        acq = (
            tx.findtext(".//transactionAmounts/transactionAcquiredDisposedCode/value")
            or ""
        )
        if code != "P" or acq != "A":
            continue

        try:
            shares = float(
                tx.findtext(".//transactionAmounts/transactionShares/value") or 0
            )
            price = float(
                tx.findtext(".//transactionAmounts/transactionPricePerShare/value") or 0
            )
        except (ValueError, TypeError):
            continue

        total_value = shares * price
        if total_value < min_value:
            continue

        tx_date = tx.findtext(".//transactionDate/value") or ""
        purchases.append(
            {
                "issuer": issuer_name,
                "ticker": ticker,
                "owner": owner_name,
                "title": title,
                "date": tx_date,
                "shares": int(shares),
                "price": round(price, 4),
                "value": round(total_value, 2),
            }
        )

    return purchases


# ── Slack formatting ──────────────────────────────────────────────────────────

def _fmt_value(v: float) -> str:
    if v >= 1_000_000:
        return f"${v / 1_000_000:.2f}M"
    return f"${v / 1_000:.0f}K"


def build_slack_message(
    purchases: list[dict], start: str, end: str, min_value: float
) -> str:
    threshold = _fmt_value(min_value)
    lines = [
        f"*:bar_chart: SEC Form 4 — Insider Purchases >{threshold}*",
        f"_Executives & Directors only | {start} → {end} UTC_",
        "",
    ]

    if not purchases:
        lines.append(
            f":white_check_mark: Scan complete — no qualifying purchases "
            f"(>{threshold}) found in this window."
        )
        return "\n".join(lines)

    lines.append(f"*{len(purchases)} purchase{'s' if len(purchases) != 1 else ''} found, ranked by value:*\n")

    shown = purchases[:TOP_N_SLACK]
    for rank, p in enumerate(shown, 1):
        label = p["ticker"] if p["ticker"] else p["issuer"]
        val_str = _fmt_value(p["value"])
        price_str = f"${p['price']:,.2f}"
        lines.append(
            f"*{rank}.* `{label}` — {p['owner']} _{p['title']}_\n"
            f"    {p['shares']:,} sh @ {price_str} = *{val_str}*   📅 {p['date']}"
        )

    if len(purchases) > TOP_N_SLACK:
        lines.append(f"\n_…and {len(purchases) - TOP_N_SLACK} more not shown_")

    lines.append(
        f"\n_Source: SEC EDGAR Form 4 filings · "
        f"https://www.sec.gov/cgi-bin/browse-edgar?action=getcurrent&type=4_"
    )
    return "\n".join(lines)


# ── Slack delivery ────────────────────────────────────────────────────────────

def post_to_slack(message: str, webhook_url: str) -> None:
    payload = json.dumps({"text": message}).encode()
    req = Request(
        webhook_url,
        data=payload,
        headers={"Content-Type": "application/json", "User-Agent": USER_AGENT},
        method="POST",
    )
    with urlopen(req, timeout=15) as resp:
        status = resp.status
        body = resp.read().decode()
    if status != 200 or body.strip() != "ok":
        raise RuntimeError(f"Slack responded {status}: {body!r}")


# ── Main ──────────────────────────────────────────────────────────────────────

def main(argv: list[str] | None = None) -> int:
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument("--hours", type=int, default=DEFAULT_HOURS,
                        help="Look-back window in hours (default: 24)")
    parser.add_argument("--min-value", type=float, default=DEFAULT_MIN_VALUE,
                        help="Minimum purchase value in USD (default: 100000)")
    parser.add_argument("--slack-webhook", default=os.getenv("SLACK_WEBHOOK_URL"),
                        help="Slack Incoming Webhook URL (or set SLACK_WEBHOOK_URL)")
    args = parser.parse_args(argv)

    now = datetime.now(timezone.utc)
    end_dt = now.strftime("%Y-%m-%d")
    start_dt = (now - timedelta(hours=args.hours)).strftime("%Y-%m-%d")

    print(
        f"[{now.strftime('%Y-%m-%dT%H:%M:%SZ')}] "
        f"Scanning Form 4 filings {start_dt} → {end_dt} "
        f"(min ${args.min_value:,.0f})",
        file=sys.stderr,
    )

    # ── 1. Search EDGAR ───────────────────────────────────────────────────────
    filings = search_form4_filings(start_dt, end_dt)
    print(f"Found {len(filings)} Form 4 filing(s) to examine", file=sys.stderr)

    # ── 2. Parse each filing ──────────────────────────────────────────────────
    all_purchases: list[dict] = []
    errors = 0

    for idx, (acc_no, cik) in enumerate(filings):
        try:
            root = get_primary_xml(acc_no, cik)
            if root is not None:
                new = parse_purchases(root, args.min_value)
                all_purchases.extend(new)
        except Exception as exc:
            errors += 1
            if errors <= 10:
                print(f"  Error {acc_no}: {exc}", file=sys.stderr)

        if idx % 50 == 0 and idx > 0:
            print(
                f"  {idx}/{len(filings)} processed — "
                f"{len(all_purchases)} qualifying purchases so far",
                file=sys.stderr,
            )

        time.sleep(INTER_REQUEST_DELAY)

    # ── 3. Rank ───────────────────────────────────────────────────────────────
    all_purchases.sort(key=lambda x: x["value"], reverse=True)
    print(
        f"Scan complete: {len(all_purchases)} qualifying purchase(s) "
        f"from {len(filings)} filing(s) ({errors} error(s))",
        file=sys.stderr,
    )

    # ── 4. Format ─────────────────────────────────────────────────────────────
    message = build_slack_message(all_purchases, start_dt, end_dt, args.min_value)
    print(message)

    # ── 5. Send to Slack ──────────────────────────────────────────────────────
    if args.slack_webhook:
        try:
            post_to_slack(message, args.slack_webhook)
            print("Posted to Slack.", file=sys.stderr)
        except Exception as exc:
            print(f"Slack delivery failed: {exc}", file=sys.stderr)
            return 1
    else:
        print(
            "(Set SLACK_WEBHOOK_URL or --slack-webhook to post to Slack)",
            file=sys.stderr,
        )

    # Also emit JSON for programmatic consumers
    json_path = "insider_purchases.json"
    with open(json_path, "w") as fh:
        json.dump(
            {"generated_at": now.isoformat(), "purchases": all_purchases},
            fh,
            indent=2,
        )
    print(f"JSON results written to {json_path}", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())
